sdks/java/harness/src/main/java/org/apache/beam/fn/harness/PrecombineGroupingTable.java

```diff
@@ -17,434 +17,392 @@
-      entry = createTableEntry(key);
-      table.put(groupingKey, entry);
-      size += PER_KEY_OVERHEAD;
-    } else {
-      size -= entry.getSize();
-    }
-    entry.add(value);
-    size += entry.getSize();
-
-    if (size >= maxSize) {
-      long targetSize = (long) (TARGET_LOAD * maxSize);
-      Iterator<GroupingTableEntry<K, InputT, AccumT>> entries = table.values().iterator();
-      while (size >= targetSize) {
-        if (!entries.hasNext()) {
-          // Should never happen, but sizes may be estimates...
-          size = 0;
-          break;
+  @VisibleForTesting
+  public void put(
+      WindowedValue<KV<K, InputT>> value, FnDataReceiver<WindowedValue<KV<K, AccumT>>> receiver)
+      throws Exception {
+    // Ignore timestamp for grouping purposes.
+    // The Pre-combine output will inherit the timestamp of one of its inputs.
+    GroupingTableKey groupingKey =
+        new GroupingTableKey(
+            value.getValue().getKey(), value.getWindows(), value.getPane(), keyCoder, keySizer);
+
+    lruMap.compute(
+        groupingKey,
+        (key, tableEntry) -> {
+          if (tableEntry == null) {
+            weight += groupingKey.getWeight();
+            tableEntry =
+                new GroupingTableEntry(
+                    groupingKey, value.getValue().getKey(), value.getValue().getValue());
+          } else {
+            weight -= tableEntry.getWeight();
+            tableEntry.add(value.getValue().getValue());
+          }
+          weight += tableEntry.getWeight();
+          return tableEntry;
+        });
+
+    // Increase the maximum only if we require it
+    maxWeight.accumulateAndGet(weight, (current, update) -> current < update ? update : current);
+
+    // Update the cache to ensure that LRU is handled appropriately and for the cache to have an
+    // opportunity to shrink the maxWeight if necessary.
+    cache.put(Key.INSTANCE, this);
+
+    // Get the updated weight now that the cache may have been shrunk and respect it
+    long currentMax = maxWeight.get();
+
+    // Only compact and output from the bundle processing thread that is inserting elements into the
+    // grouping table. This ensures that we honor the guarantee that transforms for a single bundle
+    // execute using the same thread.
+    if (weight > currentMax) {
+      // Try to compact as many the values as possible and only flush values if compaction wasn't
+      // enough.
+      for (GroupingTableEntry valueToCompact : lruMap.values()) {
+        long currentWeight = valueToCompact.getWeight();
+        valueToCompact.compact();
+        weight += valueToCompact.getWeight() - currentWeight;
+      }
+
+      if (weight > currentMax) {
+        Iterator<GroupingTableEntry> iterator = lruMap.values().iterator();
+        while (iterator.hasNext()) {
+          GroupingTableEntry valueToFlush = iterator.next();
+          weight -= valueToFlush.getWeight() + valueToFlush.getGroupingKey().getWeight();
```

Comment:
I'm having some trouble following all the different weights, and my first instinct is that since valueToFlush contains the GroupingKey, that this would count the weight of the grouping key twice (and presumably this would be bad because it wasn't counted twice when being originally added to the max weight).